File: concepts/graphs/undirected/cycle_dfs.cpp
#include <bits/stdc++.h>
using namespace std;

class Solution {
   public:
    vector<bool> visited;
    bool isCycleDFS(vector<int> adj[], int current, int parent) {
        // mark node visited
        visited[current] = true;

        // from that current node visit next neighbour if its not visited
        for (int &v : adj[current]) {
            // skip parent position since we came from that point
            // if we go back it will be considered as cycle even though its not
            if (v == parent) {
                continue;
            }

            if (visited[v]) {
                // we found cycle since node is visited twice
                return true;
            }

            if (isCycleDFS(adj, v, current)) {
                return true;
            }
        }

        return false;
    }
    bool isCycle(int V, vector<int> adj[]) {
        visited.resize(V, false);

        // we are assuming graphs are component bases thus this loop is required
        for (int i = 0; i < V; i++) {
            // when node is not visited
            if (!visited[i]) {
                if (isCycleDFS(adj, i, -1)) {
                    return true;
                }
            }
        }
        return false;
    }
};

File: concepts/graphs/undirected/cycle_dfs.py
from typing import List
from collections import defaultdict


class Solution:
    def __init__(self) -> None:
        self.visited = defaultdict(bool)
        
    def isCycleDFS(self, adj, curr, parent):
        self.visited[curr] = True
  
        for child in adj[curr]:
            if (child == parent):
                continue

            if (self.visited[child]):
                return True
            if (self.isCycleDFS(adj=adj, curr=child, parent=curr)):
                return True
        return False

    def isCycle(self, V: int, adj: List[List[int]]) -> bool:
        for i in range(0, V):
            if not self.visited[i] and self.isCycleDFS(adj, i, -1):
                return True
        return False


ans = Solution().isCycle(V=5, adj=[[1], [0, 2, 4], [1, 3], [2, 4], [1, 3]])
print(ans)
